test/functional-portable/kubernetes/noncloud/deploymenttemplate_test.go

```diff
@@ -0,0 +1,358 @@
+
+	template, err := os.ReadFile(templateFilePath)
+	require.NoError(t, err)
+
+	// Create the namespace, if it already exists we can ignore the error.
+	_, err = opts.K8sClient.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}, metav1.CreateOptions{})
+	require.NoError(t, controller_runtime.IgnoreAlreadyExists(err))
+
+	deploymentTemplate := makeDeploymentTemplate(types.NamespacedName{Name: name, Namespace: namespace}, string(template), providerConfig, parametersMap)
+
+	t.Run("Create DeploymentTemplate", func(t *testing.T) {
+		t.Log("Creating DeploymentTemplate")
+		err = opts.Client.Create(ctx, deploymentTemplate)
+		require.NoError(t, err)
+	})
+
+	t.Run("Check DeploymentTemplate status", func(t *testing.T) {
+		ctx, cancel := testcontext.NewWithCancel(t)
+		defer cancel()
+
+		// Get resource version
+		err = opts.Client.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, deploymentTemplate)
+		require.NoError(t, err)
+
+		t.Log("Waiting for DeploymentTemplate ready")
+		deploymentTemplate, err := waitForDeploymentTemplateReady(t, ctx, types.NamespacedName{Name: name, Namespace: namespace}, opts.Client, deploymentTemplate.ResourceVersion)
+		require.NoError(t, err)
+
+		scope, err := reconciler.ParseDeploymentScopeFromProviderConfig(deploymentTemplate.Spec.ProviderConfig)
+		require.NoError(t, err)
+
+		expectedResources := [][]string{
+			{"Applications.Core/environments", fmt.Sprintf("%s-env", name)},
+			{"Applications.Core/applications", fmt.Sprintf("%s-app", name)},
+			{"Applications.Datastores/redisCaches", fmt.Sprintf("%s-recipe", name)},
+		}
+
+		assertExpectedResourcesExist(t, ctx, scope, expectedResources, opts.Connection)
+	})
+
+	t.Run("Delete DeploymentTemplate", func(t *testing.T) {
+		t.Log("Deleting DeploymentTemplate")
+		err = opts.Client.Delete(ctx, deploymentTemplate)
+		require.NoError(t, err)
+
+		require.Eventually(t, func() bool {
+			err = opts.Client.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, deploymentTemplate)
+			return apierrors.IsNotFound(err)
+		}, time.Minute*3, time.Second*5, "waiting for deploymentTemplate to be deleted")
+	})
+}
+
+func makeDeploymentTemplate(name types.NamespacedName, template, providerConfig string, parameters map[string]string) *radappiov1alpha3.DeploymentTemplate {
+	deploymentTemplate := &radappiov1alpha3.DeploymentTemplate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+		Spec: radappiov1alpha3.DeploymentTemplateSpec{
+			Template:       template,
+			Parameters:     parameters,
+			ProviderConfig: providerConfig,
+		},
+	}
+
+	return deploymentTemplate
+}
+
+func waitForDeploymentTemplateReady(t *testing.T, ctx context.Context, name types.NamespacedName, client controller_runtime.WithWatch, initialVersion string) (*radappiov1alpha3.DeploymentTemplate, error) {
+	// Based on https://gist.github.com/PrasadG193/52faed6499d2ec739f9630b9d044ffdc
```

Comment:
Can you add comments for what this code does

Reply:
done